src/server/array_func.c

```diff
@@ -653,32 +652,17 @@ setup_arrayjob_attrs(attribute *pattr, void *pobj, int mode)
 	pjob->ji_wattr[(int)JOB_ATR_array].at_flags = ATR_VFLAG_SET | ATR_VFLAG_MODCACHE;
 	job_attr_def[(int)JOB_ATR_array_state_count].at_free(&pjob->ji_wattr[(int)JOB_ATR_array_state_count]);
 
-	if (mode == ATR_ACTION_RECOV) {
-		int x, y, z, ct;
-		char *ep;
-
-		/* on recovery ... */
-		/* parse the various components again, since we dont store them */
-		if (parse_subjob_index(pattr->at_val.at_str, &ep, &x, &y, &z, &ct) != 0)
+	if ((mode == ATR_ACTION_NEW) || (mode == ATR_ACTION_RECOV)) {
+		if (pjob->ji_ajtrk)
+			free(pjob->ji_ajtrk);
+		if ((pjob->ji_ajtrk = mk_subjob_index_tbl(pjob->ji_wattr[(int)JOB_ATR_array_indices_submitted].at_val.at_str, JOB_STATE_QUEUED)) == NULL)
 			return PBSE_BADATVAL;
+	}
 
-		pjob->ji_ajtrk->tkm_ct = ct;
-		pjob->ji_ajtrk->tkm_step = z;
-		pjob->ji_ajtrk->tkm_flags = 0;
-
-		/* reset counts and any running/exiting subjob to queued */
-		for (i=0; i < PBS_NUMJOBSTATE; ++i)
-			pjob->ji_ajtrk->tkm_subjsct[i] = 0;
-		for (i=0; i < pjob->ji_ajtrk->tkm_ct; ++i) {
-			if ((pjob->ji_ajtrk->tkm_tbl[i].trk_status == JOB_STATE_RUNNING)
-				|| (pjob->ji_ajtrk->tkm_tbl[i].trk_status == JOB_STATE_EXITING))
-				pjob->ji_ajtrk->tkm_tbl[i].trk_status =JOB_STATE_QUEUED;
-			pjob->ji_ajtrk->tkm_subjsct[pjob->ji_ajtrk->tkm_tbl[i].trk_status]++;
-		}
-
-		/* clear and reset array_indices_remaining to new value */
-		job_attr_def[(int)JOB_ATR_array_indices_remaining].at_free(&pjob->ji_wattr[(int)JOB_ATR_array_indices_remaining]);
-		job_attr_def[(int)JOB_ATR_array_indices_remaining].at_decode(&pjob->ji_wattr[(int)JOB_ATR_array_indices_remaining], NULL, NULL, cvt_range(pjob->ji_ajtrk, JOB_STATE_QUEUED));
+	if (mode == ATR_ACTION_RECOV) {
+		/* set flags in attribute so stat_job will update the attr string */
+		pjob->ji_wattr[(int)JOB_ATR_array_indices_remaining].at_flags |=
```

Comment:
As discussed where is the attribute value for JOB_ATR_array_indices_remaining being set. Here you are directly setting the flags.

Reply:
Yes the value and the flag should be set in an "atomic" operation. The right thing would be do call the set function along with decode.

Reply:
*JOB_ATR_array_indices_remaining* kind of stands out wrt other JOB_ATR_xxx atribs operations. I think this is done to speed up things for job arrays. *JOB_ATR_array_indices_remaining* is only set in status_job() here https://github.com/PBSPro/pbspro/blob/master/src/server/stat_job.c#L266-L272  and is marked for updation using the ATR_VFLAG_MODCACHE at various places.  So "atomic" operation is not possible for *JOB_ATR_array_indices_remaining*.

Reply:
As discussed if status_job() is not called for whatever reasons then JOB_ATR_array_indices_remaining data is not consistent. May be a time task which updates this attribute helps here just a thought.

Reply:
Yes, This is due to flaw in our current design on how we are recalculating JOB_ATR_array_indices_remaining. This will deserve a story/RFE of its own. Till then we can feel safe by the fact that status_job() gets invoked for each sched cycle. I have noted this down for in my post merge checklist

Reply:
After reading the code carefully, i figured what you have done is correct. The flag is not about setting the value, but since data was loaded from the database, you are setting the MODCACHE flag such that any future call to the stat API would re-encode the value of this attribute. This looks fine to me.